File: Leetcode/2483.py
class Solution:
    def bestClosingTime(self, customers: str) -> int:
        n: int = len(customers)
        arr: list[int] = [int(i == "Y") for i in customers]
        tot = sum(arr)
        min_penalty = tot
        ans = 0

        for i in range(1, n):
            arr[i] += arr[i - 1]
            curr_penalty = (i - arr[i - 1]) + (tot - arr[i - 1])
            if curr_penalty < min_penalty:
                min_penalty = curr_penalty
                ans = i

        return ans if min_penalty <= (n - tot) else n


def func():
    obj = Solution()

    s = "YYNY"

    s = "NNNYNN"

    ans = obj.bestClosingTime(s)

    print(ans)


if __name__ == "__main__":
    func()

File: Leetcode/2483.go
package Leetcode

import "strings"

func bestClosingTime(customers string) int {
	n := len(customers)
	total_customers := strings.Count(customers, "Y")
	min_penalty := total_customers
	ans := 0
	current_customers := 0
	if customers[0] == 'Y' {
		current_customers = 1
	}

	for i := 1; i < n; i++ {
		curr_penalty := (i - current_customers) + (total_customers - current_customers)
		if curr_penalty < min_penalty {
			min_penalty = curr_penalty
			ans = i
		}
		if customers[i] == 'Y' {
			current_customers++
		}
	}

	if min_penalty <= (n - total_customers) {
		return ans
	}
	return n

}
